modmail/config.py

```diff
@@ -1,144 +1,648 @@
-    prefix: str = "?"
-    token: str = None
 
-    class Config:
-        # env_prefix = "bot."
-        allow_mutation = False
+def load_env(env_file: os.PathLike = None, existing_cfg_dict: dict = None) -> dict:
+    """
+    Load a configuration dictionary from the specified env file and environment variables.
+
+    All dependencies for this will always be installed.
+    """
+    if env_file is None:
+        env_file = _CWD / ".env"
+    else:
+        env_file = pathlib.Path(env_file)
 
+    if not existing_cfg_dict:
+        existing_cfg_dict = defaultdict(_generate_default_dict)
 
-class BotMode(BaseSettings):
+    existing_cfg_dict = attr.asdict(_build_class(Cfg, dotenv_file=env_file, defaults=existing_cfg_dict))
+
+    return existing_cfg_dict
+
+
+def load_toml(path: os.PathLike = None) -> defaultdict:
     """
-    Bot mode.
+    Load a configuration dictionary from the specified toml file.
 
-    Used to determine when the bot will run.
+    All dependencies for this will always be installed.
     """
+    if path is None:
+        path = (_CWD / (USER_CONFIG_FILE_NAME + ".toml"),)
+    else:
+        # fully resolve path
+        path = pathlib.Path(path)
 
-    production: bool = True
-    plugin_dev: bool = False
-    develop: bool = False
+    if not path.is_file():
+        raise CfgLoadError("The provided toml file path is not a valid file.")
 
+    try:
+        with open(path) as f:
+            return defaultdict(lambda: marshmallow.missing, atoml.parse(f.read()).value)
+    except Exception as e:
+        raise CfgLoadError from e
 
-class Colors(BaseSettings):
+
+def load_yaml(path: os.PathLike) -> dict:
     """
-    Default colors.
+    Load a configuration dictionary from the specified yaml file.
 
-    These should only be changed here to change the default colors.
+    The dependency for this may not be installed, as toml is already used elsewhere, so this is optional.
+
+    In order to keep errors at a minimum, this function checks if both pyyaml is installed
+    and if a yaml configuration file exists before raising an exception.
     """
+    if path is None:
+        path = (_CWD / (USER_CONFIG_FILE_NAME + ".yaml"),)
+    else:
+        path = pathlib.Path(path)
+
+    states = [
+        ("The yaml library is not installed.", yaml is not None),
+        ("The provided yaml config path does not exist.", path.exists()),
+        ("The provided yaml config file is not a readable file.", path.is_file()),
+    ]
+    if errors := "\n".join(msg for msg, check in states if not check):
+        raise CfgLoadError(errors)
+
+    try:
+        with open(path, "r") as f:
+            return defaultdict(lambda: marshmallow.missing, yaml.load(f.read(), Loader=yaml.SafeLoader))
+    except Exception as e:
+        raise CfgLoadError from e
 
-    embed_color: ColorBase = "0087BD"
 
+DictT = typing.TypeVar("DictT", bound=typing.Dict[str, typing.Any])
+
+
+def _remove_extra_values(klass: type, dit: DictT) -> DictT:
```

Comment:
`dit` looks too much like a typo

Reply:
We should also have an option here to warn when this deletes things, for debugging purposes

Reply:
52a339d2799b84c2edd61883ade1c75207fca332